src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,161 @@
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws CopycatException {
+    try {
+      acquireLease();
+      WALEntry key = new WALEntry(tempFile);
+      WALEntry value = new WALEntry(committedFile);
+      writer.append(key, value);
+      writer.hsync();
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private void acquireLease() throws CopycatException {
+    long initialSleepIntervalMillis = sleepIntervalMillis;
+    while (sleepIntervalMillis < MAX_SLEEP_INERVAL_MILLIS) {
+      try {
+        if (writer == null) {
+          writer = WALFile.createWriter(conf,
+                                        Writer.file(new Path(logFile)),
+                                        Writer.appendIfExists(true));
+        }
+        log.info("Successfully acquired lease for {}", logFile);
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMillis = sleepIntervalMillis * 2;
+          log.info("Cannot acquire lease on WAL {}", logFile);
+          try {
+            Thread.sleep(sleepIntervalMillis);
+          } catch (InterruptedException ie) {
+            throw new CopycatException(ie);
+          }
+        } else {
+          throw new CopycatException(e);
+        }
+      } catch (IOException e) {
+        throw new CopycatException("Error creating writer for log file " + logFile, e);
+      }
+    }
+    if (sleepIntervalMillis >= MAX_SLEEP_INERVAL_MILLIS) {
+      if (num_retry <= MAX_NUM_RETRY) {
+        num_retry++;
+        throw new TransientException("Cannot acquire lease after timeout, will retry.");
+      } else {
+        throw new CopycatException("Cannot acquire lease after retry.");
+      }
+    }
+    sleepIntervalMillis = initialSleepIntervalMillis;
+  }
+
+  public void apply() throws CopycatException {
+    try {
+      if (!fs.exists(new Path(logFile))) {
+        return;
+      }
+      acquireLease();
+      if (reader == null) {
+        reader = new WALFile.Reader(conf, Reader.file(new Path(logFile)));
+      }
+      WALEntry key = new WALEntry();
+      WALEntry value = new WALEntry();
+      while (reader.next(key, value)) {
+        String tempFile = key.getFilename();
+        String committedFile = value.getFilename();
+        Path tempPath = new Path(tempFile);
+        if (fs.exists(tempPath)) {
+          FileUtils.renameFile(fs, tempFile, committedFile);
+        }
+      }
+      truncate();
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private void truncate() throws CopycatException {
+    try {
+      String oldLogFile = logFile + ".1";
+      fs.delete(new Path(oldLogFile), false);
+      FileUtils.renameFile(fs, logFile, oldLogFile);
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  public void close() throws CopycatException {
+    try {
+      if (writer != null) {
```

Comment:
Similar comment as I mentioned earlier -- if there's an exception, do we want to cancel the entire cleanup process? Should these be in separate try/catch blocks? We might not need any change here, but I'd like to think through the implications of, e.g., the very first method call failing and what the fallout is.